MCO-1437: MCO-1476: MCO-1477: MCO-1284: Adapt MCO to OCL v1 API

djoshy opened the pull request:
This PR does the following:
- bumps client-go to capture OCL v1 API helpers
- reconciles machine-os-builder and the operator to handle OCL v1 API
- reconciles existing units/e2es to handle OCL v1 API
- adds a global pull secret cloning mechanism since `BaseImagePullSecret` is optional in v1 API

How to test/verify:
 - TechPreview should be enabled
 - All existing units/e2es should pass
 - OCL workflows should be tested with v1 CRs(v1alpha1 CRDs will no longer work).
 

**Note**: This PR should be tested and merged along with https://github.com/openshift/api/pull/2134 

Clusterbot incantation: 
```
launch openshift/machine-config-operator#4756,openshift/api#2134 techpreview
```

djoshy commented on pkg/controller/build/buildrequest/buildrequest.go:
```diff
@@ -525,7 +518,7 @@ func (br buildRequestImpl) toBuildahPod() *corev1.Pod {
 					Name:            "wait-for-done",
 					Command:         append(command, waitScript),
-					Image:           br.opts.getBaseOSImagePullspec(),
+					Image:           br.opts.getBaseOSImagePullspec(), // QOCL: What do we replace this with?
```
I'm guessing this would be our MC's base image?

cheesesashimi replied:
Correct. It's the value from `b.OSImageURLConfig.BaseOSContainerImage`.

djoshy replied:
ack, will update!

djoshy commented on pkg/controller/build/buildrequest/machineosbuild.go:
```diff
@@ -63,20 +62,14 @@ func (m *MachineOSBuildOpts) validateForHash() error {
-	moscCopy.Spec.BuildInputs.BaseOSImagePullspec = ""
-	moscCopy.Spec.BuildInputs.BaseOSExtensionsImagePullspec = ""
-	moscCopy.Spec.BuildInputs.ReleaseVersion = ""
+	// QOCL:Does this function do anything anymore?
```
These fields have been removed, unsure about this function's usage

cheesesashimi replied:
This function can probably go away.

djoshy commented on pkg/controller/build/buildrequest/machineosbuild.go:
```diff
@@ -227,18 +211,17 @@ func NewMachineOSBuild(opts MachineOSBuildOpts) (*mcfgv1alpha1.MachineOSBuild, e
 			},
 		},
-		Status: mcfgv1alpha1.MachineOSBuildStatus{
+		//QOCL:Was this type change to pointer for status intended?
```
Just a double check

djoshy replied:
chatted with Jerry and confirmed this is intended.

djoshy commented on pkg/controller/build/buildrequest/machineosbuild_test.go:
```diff
@@ -81,6 +80,7 @@ func TestMachineOSBuild(t *testing.T) {
 				OSImageURLConfig:  fixtures.OSImageURLConfig(),
 			},
 		},
+		/* QOCL: Are these test cases valid anymore?
```
There are 6 test cases here that seem to exercise fields that have been removed: `BaseOSExtensionsImagePullspe`, `BaseOSImagePullspec` and `ReleaseVersion`

cheesesashimi replied:
In my opinion, we still need those fields in order to produce a hashed name for a MachineOSBuild. The difference is that we only have a single canonical source for these values with the new API changes.

djoshy commented on test/e2e-ocl/onclusterlayering_test.go:
```diff
@@ -319,65 +318,71 @@ func TestMachineConfigPoolChangeRestartsBuild(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	t.Cleanup(cancel)
+	/*
+		QOCL: Need a new failure mode for this as BaseImagePullspec field doesn't exist anymore
```
With the removal of `BaseImagePullSpec`, I'm not sure the test can be initiated in the same way here, so I'm open to ideas here

cheesesashimi replied:
We can do a Containerfile with invalid syntax instead. It will just take a bit longer to produce the failure.

djoshy commented on test/e2e-ocl/onclusterlayering_test.go:
```diff
@@ -572,7 +577,8 @@ func TestMCDGetsMachineOSConfigSecrets(t *testing.T) {
 	// Assign the secret name to the MachineOSConfig.
-	mosc.Spec.BuildOutputs.CurrentImagePullSecret.Name = secretName
+	// QOCL: This field doesn't exist anymore, is this test valid?
+	//mosc.Spec.CurrentImagePullSecret.Name = secretName
```
Another scenario where I believe the test currently depends on this field to run correctly

cheesesashimi replied:
This whole test can be dropped in favor of a test that just validates whether the global image pull secret is written to the nodes' filesystem.

djoshy replied:
I believe we already do this as part of the internal registry secret cert writer path, where the global pull secret +  internal registry secret is written to disk at `/etc/mco/internal-registry-pull-secret.json` : https://github.com/openshift/machine-config-operator/pull/4395/commits/2d77dcf4a9b8817f8e58026e31346056f22570f8

djoshy commented on test/helpers/machineosconfigbuilder.go:
```diff
@@ -21,81 +21,58 @@ func NewMachineOSConfigBuilder(name string) *MachineOSConfigBuilder {
 	return m
 }
 
+// QOCL:Is CurrentImagePullSecret gone? It wasn't listed in the story as one of the fields that was removed.
```
in ref to https://github.com/openshift/machine-config-operator/pull/4756/commits/6064540dc3a033132a27ae975bc44efb67588e35#r1884535019

cheesesashimi replied:
I'm pretty sure it is. This value will now come from the global image pull secret instead.

djoshy replied:
ack, updated https://issues.redhat.com/browse/MCO-1476 to reflect this

cheesesashimi commented on pkg/controller/build/reconciler.go:
```diff
@@ -101,7 +100,8 @@ func (b *buildReconciler) updateMachineOSConfig(ctx context.Context, old, cur *m
 
 	// Whenever the build inputs have changed, create a new MachineOSBuild.
-	if !equality.Semantic.DeepEqual(old.Spec.BuildInputs, cur.Spec.BuildInputs) {
+	// QOCL: ok to compare the spec.MachineConfigPool by comparing the whole spec here? Inputs didnt include the pool before.
```
It's OK in my opinion. What I don't know is whether the MCP reference is immutable. In my opinion, it should be.

cheesesashimi commented on pkg/controller/build/reconciler.go:
```diff
@@ -302,23 +302,23 @@ func (b *buildReconciler) updateMachineOSConfigStatus(ctx context.Context, mosc
-	// TODO: Reconsider this.
-	mosc.Status.ObservedGeneration += mosc.GetGeneration()
+	mosc.Status.CurrentImagePullSpec = mosb.Status.DigestedImagePushSpec
+	// QOCL: Double check that this is the only place the controller updates the MOSC statuses, in response to a change in MOSC spec.
```
This should be the only place, but double-check just to be safe.

cheesesashimi commented on pkg/operator/render_test.go:
```diff
@@ -306,6 +305,7 @@ func TestRenderAsset(t *testing.T) {
 			},
 		},
 		// Tests that the MCD DaemonSet gets MachineOSConfig secrets mounted into it.
+		/*QOCL: Another instance of what are we doing with CurrentImagePullSecret?
```
To simplify things, we can probably remove this test while also removing the volume mounts from the MCD manifest since this is no longer needed.

djoshy replied:
I believe you're talking about the internal registry path [I mentioned here](https://github.com/openshift/machine-config-operator/pull/4756#discussion_r1887054249) and I think we already have something that [directs rpm-ostree to use the pull secret](https://github.com/openshift/machine-config-operator/blob/36e2d57b1583811e924ed91452b4b3df083b6931/pkg/daemon/rpm-ostree.go#L180) if that file exists. Correct me if I'm missing something here!

djoshy replied:
I've pulled out the code specific to the secret mounting and split it off into a separate commit. I've also pulled out all related tests for this particular path - the render unit test mentioned in this thread and the e2e test mentioned here https://github.com/openshift/machine-config-operator/pull/4756#discussion_r1884564248 

I do agree that it is a good idea to add a test to ensure that global pull secret is written to disk. I think we might have some units for this path on the operator and the daemon level already, but not an e2e - I could be wrong though. Do you think we should add that work to this PR, or open up a new card for it?

djoshy replied:
I was about to open a spike to explore this, but thinking a bit more about it, the `kubeletAuthFile` is written down via a MachineConfig, and its contents are templated out from ControllerConfig. The `kubeletAuthFile` is in the list of no reboot node disruption policies, so it wouldn't disruptive if an update happened to it. Wouldn't this be an e2e to test out a MachineConfig file write? :thinking: I am overthinking this? :smile:

yuqi-zhang replied:
Generally speaking, file and cert writing should be covered. I guess the question is whether we'd need to have something to "ensure MCD has access to the right registries"?

I'm not really sure that's needed. We can move content around to de-duplicate, but regardless, let's do that as a followup

djoshy commented:
/test unit

djoshy commented:
/test unit
/test verify

djoshy commented on pkg/controller/build/imagebuilder/base.go:
```diff
@@ -278,35 +278,35 @@ type kubeObject interface {
-		PodImageBuilder: &mcfgv1alpha1.ObjectReference{
+		Job: &mcfgv1.ObjectReference{
 			Name:      obj.GetName(),
-			Group:     obj.GroupVersionKind().Group,
+			Group:     "tbd-group",
 			Namespace: obj.GetNamespace(),
-			Resource:  obj.GetResourceVersion(),
+			Resource:  "tbd-resource",
```
The values being applied fail API validation. The API requires that all these fields should validate against `format.dns1123Subdomain()`. In my testing Group seemed to be often assigned an empty value. Resource Version, which is normally a number, being assigned to Resource also seems like a mistake. Open to suggestions here!

cheesesashimi replied:
thought: I would say that we could remove the group and resource values since they're not really used anywhere else, but the API validation requires them. I feel like we can and should loosen the validation rules and optionality.

yuqi-zhang replied:
I'm in favour of making them optional. Looks like they were made required in the alpha API so I didn't modify those. I'm not 100% sure if the API allows for a required field to go back to optional (maybe it's fine since it hasn't been used yet?) If that's not available, what validation rules would we like to see?

djoshy replied:
+1 to making them optional - once we address Jerry's questions above, we'll need the API to be updated and brought back in. I'll leave the placeholders in for the moment until we have a resolution.

cheesesashimi replied:
For validation rules, I'm not sure what makes sense here.

djoshy replied:
Did some more digging here as to why these fields were coming up empty, it's because we are creating and using the job object before passing it through a client. The GVK values won't be populated in such a case. We could just manually set these instead. This approach would mean no API changes, but it also means that we'll be accepting the API in its current state.

yuqi-zhang replied:
Let's set up the API change anyways, and if the API team rejects it, we can always fall back to just having the fields properly

djoshy commented on pkg/operator/sync.go:
```diff
@@ -1463,6 +1468,58 @@ func (optr *Operator) reconcileSimpleContentAccessSecrets(layeredMCPs []*mcfgv1.
+	// QOCL: Do we want to fatally exit here?
+	globalPullSecret, err := optr.ocSecretLister.Secrets(ctrlcommon.OpenshiftConfigNamespace).Get("pull-secret")
+	if err != nil {
+		return fmt.Errorf("error fetching cluster pull secret: %w", err)
```
Do we want to fatally exit here? If this secret is missing in cluster, the build will certainly fail, I'd think.

umohnani8 replied:
Yes, the build will most definitely fail without the global pull secret. Makes sense to fatally exit here to me

djoshy commented on pkg/operator/sync.go:
```diff
@@ -1463,6 +1468,58 @@ func (optr *Operator) reconcileSimpleContentAccessSecrets(layeredMCPs []*mcfgv1.
+	// If it does exist, check if an update is required before making the update call.
+	if !reflect.DeepEqual(currentSecretCopy.Data, globalPullSecret.Data) {
+		klog.Infof("updating %s", ctrlcommon.GlobalPullSecretCopyName)
+		_, err := optr.kubeClient.CoreV1().Secrets(ctrlcommon.MCONamespace).Update(context.TODO(), globalPullSecretCopy, metav1.UpdateOptions{})
```
Not sure if this update mechanism is necessary, I don't know if the global secret will ever get updated after installation?

umohnani8 replied:
Looking through the docs, looks like the global pull secret can be updated for certain scenarios. I think having this update mechanism makes sense

djoshy commented on pkg/controller/build/osbuildcontroller_test.go:
```diff
@@ -341,7 +341,9 @@ func TestOSBuildControllerReusesPreviouslyBuiltImage(t *testing.T) {
-
+/*
+QOCL: This test seems to pass individually, but fails when run with other unit tests. Other tests in this package
+// seems to exhibit similar failures but at a lower rate. Possible object conflicts/worth exploring as a new bug?
```
These tests seemed to be conflicting with other tests for some reason, would appreciate some pointers if I'm missing something!

djoshy replied:
Gotcha! Would you be open to me spinning this off and investigating them as a separate bug?

cheesesashimi replied:
To be clear, when you say "excluded", do you mean just skipping `TestOSBuildController`? Or do you mean excluding it from parallelization?

djoshy replied:
Yes, skipping or commenting out the test for the time being. The `TestOSBuildController` failures seem to happen regardless of how I applied the parallelization. I do think the parallelization is one of the issues here, but this particular test seems to be quite a bit more flakier than the rest(even in serial mode) and it may be worth exploring deeper.

yuqi-zhang replied:
I am :+1: on making a separate bug for this. When you say "even lower rate" is it around 10%? 1%? Just to see if we 'd need to exclude the test for stability purposes until we can resolve the fundamental issue.

djoshy replied:
Reran these; 1000 count each and failfast off, so we get comparable data:

Parallelized with `TestOSBuildController` included:  2.5% fail rate
Parallelized with `TestOSBuildController` not included: 0.24% fail rate(current implementation)
Serial with `TestOSBuildController` included: 0.32% fail rate
Serial with `TestOSBuildController` not included: 0% fail rate

djoshy replied:
Discussed with Jerry, we're opting to go with the third option, serial with the `TestOSBuildController` included. Will open a low prio bug in the backlog to investigate this test in parallel mode/

djoshy commented on pkg/operator/sync.go:
```diff
@@ -1417,10 +1416,26 @@ func (optr *Operator) reconcileSimpleContentAccessSecrets(layeredMCPs []*mcfgv1.
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      secretName,
 				Namespace: ctrlcommon.MCONamespace,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: currentPool.APIVersion,
+						Kind:       currentPool.Kind,
+						Name:       currentPool.ObjectMeta.Name,
+						UID:        currentPool.ObjectMeta.UID,
+					},
+				},
```
This was a minor fix for a particular scenario I encountered during e2es: If the pool was deleted prior to being opted out of layering, this sync function wouldn't delete it. Adding the MCP as an owner field seemed like the cleanest fix.

djoshy commented on test/e2e-ocl/onclusterlayering_test.go:
```diff
@@ -313,75 +313,67 @@ func TestMachineConfigPoolChangeRestartsBuild(t *testing.T) {
+// an invalid containerfile. After failure, it edits the  MachineOSConfig
+// with the expectation that the failed build and its  will be deleted and a new
+// build will start in its place.
 func TestGracefulBuildFailureRecovery(t *testing.T) {
```
This test has been updated to use a bad container file as suggested in https://github.com/openshift/machine-config-operator/pull/4756#discussion_r1884562915. It now takes about 17 minutes to run locally, not sure how fast it used to be prior to that!

djoshy commented on test/e2e-ocl/onclusterlayering_test.go:
```diff
@@ -811,14 +803,11 @@ func prepareForOnClusterLayeringTest(t *testing.T, cs *framework.ClientSet, test
 
 	imagestreamObjMeta := metav1.ObjectMeta{
-		Name:      "os-image",
-		Namespace: strings.ToLower(t.Name()),
+		Name: "os-image",
 	}
```
This was necessary because we are no longer able to inject a custom pull secret. By placing the image in the MCO's slice of the registry, the daemon will be able to handle this rebase with existing secrets.

djoshy commented:
This PR has been rebased on the latest forks of the API and client-go from Jerry, so it is fairly close to the final shape of the final API. I've left comments above so hopefully we can remember where we left off before the break :christmas_tree: 

Since we can't run the e2es until the v1 CRDs have landed, I took the liberty of testing the e2es locally and we seemed to be only failing on `TestYumReposBuilds`. I am unsure if running in CI will fare better results, but just thought I'd mention it in case something I did here did break them! :smile:

djoshy commented:
/test unit
/test verify

djoshy commented:
Updated to include the true API/client-go since https://github.com/openshift/api/pull/2090 and https://github.com/openshift/client-go/pull/305 have landed. 

/test unit
/test verify

opened a draft PR to pull in the CRD, so going to try this:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/test unit
/test verify

I removed the API bump, since it has landed as part of https://github.com/openshift/machine-config-operator/pull/4772

yuqi-zhang commented on devex/cmd/onclustertesting/machineosconfigs.go:
```diff
@@ -25,47 +24,41 @@ type moscOpts struct {
-			BuildOutputs: mcfgv1alpha1.BuildOutputs{
-				CurrentImagePullSecret: mcfgv1alpha1.ImageSecretObjectReference{
-					Name: opts.finalPullSecretName,
+			RenderedImagePushSecret: mcfgv1.ImageSecretObjectReference{
```
Curious, was this field not already populated to disk? Or was this being processed via one of the removed fields?

yuqi-zhang commented on pkg/controller/build/buildrequest/buildrequestopts.go:
```diff
@@ -44,36 +45,6 @@ type BuildRequestOpts struct { //nolint:revive // This name is fine.
 	AdditionalTrustBundle []byte
 }
 
-// Gets the extensions image pullspec from the MachineOSConfig if available.
```
I vaguely recall we opted to remove this because we didn't really want the user to specify an extensions image not shipped with that version of OCP, if the user wanted to leverage the extensions. Curious what the use case would be to allow for it

cheesesashimi replied:
My thought was handling upgrades to allow someone to prebuild the OS images in advance. Though we probably don't need to expose that via the API.

djoshy commented on pkg/operator/render.go:
```diff
@@ -42,7 +41,7 @@ type renderConfig struct {
 	Constants              map[string]string
 	PointerConfig          string
-	MachineOSConfigs       []*mcfgv1alpha1.MachineOSConfig
+	MachineOSConfigs       []*mcfgv1.MachineOSConfig
```
Yep, will remove this! Would you mind addressing the question here? https://github.com/openshift/machine-config-operator/pull/4756#discussion_r1887547123 It's an older thread that is somewhat related.

cheesesashimi replied:
Addressed in that thread.

djoshy commented on pkg/operator/sync.go:
```diff
@@ -1448,6 +1468,58 @@ func (optr *Operator) reconcileSimpleContentAccessSecrets(layeredMCPs []*mcfgv1.
 
+func (optr *Operator) reconcileGlobalPullSecretCopy(layeredMCPs []*mcfgv1.MachineConfigPool) error {
+	secretCopyExists := true
+	currentSecretCopy, err := optr.mcoSecretLister.Secrets(ctrlcommon.MCONamespace).Get(ctrlcommon.GlobalPullSecretCopyName)
```
I can definitely explore doing this! I don't think BuildController has access to `controllerconfig` objects at the moment, so it would need an RBAC change. I did this approach because we have a similar flow for pulling in the [etc-pki-entitlement secret](https://github.com/openshift/machine-config-operator/pull/4613). Let me know if you feel strongly about this and I can try to implement it.

cheesesashimi replied:
No worries! I forgot that it had ControllerConfig access as well since there are some things it pulls from the MCO's ConfigMaps which can also be found there.

djoshy replied:
Discussed with Zack on slack and the controllerconfig only has an object reference and not the actual contents of the secret itself, so we will be keeping the existing approach here.

djoshy commented:
In this last push(s):
- removed these [units tests](https://github.com/openshift/machine-config-operator/pull/4756#discussion_r1906030531)
- made `HasUserDefinedBaseImagePullSecret` private
- removed MachineOSConfigs from [render fields list](https://github.com/openshift/machine-config-operator/pull/4756#discussion_r1905873854)
- removed rebuild portion of the TestOnClusterLayering e2e as [suggested in bug](https://issues.redhat.com/browse/OCPBUGS-46421)

/test unit
/test verify
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
https://prow.ci.openshift.org/view/gs/test-platform-results/logs/multi-pr-openshift-machine-config-operator-4756-openshift-machine-config-operator-4756-openshift-api-2134-e2e-gcp-op-ocl/1877009040354054144 is green :partying_face:

cheesesashimi commented on pkg/daemon/certificate_writer.go:
```diff
@@ -371,12 +358,7 @@ func (dn *Daemon) syncOSImagePullSecrets(controllerConfig *mcfgv1.ControllerConf
-	}
-
-	if err := writeToDisk(map[string][]byte{imageRegistryAuthFile: merged}); err != nil {
+	if err := writeToDisk(map[string][]byte{imageRegistryAuthFile: controllerConfig.Spec.InternalRegistryPullSecret}); err != nil {
```
You're right about the confusion since it definitely got me here! I completely overlooked where this was occurring. Please ignore my initial comment here about needing to use the global image pull secret. Instead, I think we can probably remove this function (`syncClusterPullSecrets()`) entirely. Here's why:

This function was added to merge the contents of the mounted image pull secret (originally provided by the cluster admin via the MachineOSConfig) with the contents of the internal registry image pull secret. Since the cluster admin now must add the final OS image pull secret to the global image pull secret, this path doesn't make sense anymore.

What I'm not sure about is whether anything else uses the internal image registry secrets within the MCD. My gut says no, but I could be wrong. Therefore, I propose that we rename `syncClusterPullSecrets()` to something like `syncInternalRegistryPullSecrets()` and rename `imageRegistryAuthFile` to `internalRegistryAuthFile`. We can also drop the `osImagePullSecretDir` constant since it's no longer being used. :smile:

djoshy replied:
No worries, I agree with the name changes you've suggested, that makes things a lot more clear! :smile: 
> What I'm not sure about is whether anything else uses the internal image registry secrets within the MCD. My gut says no, but I could be wrong. 

So just stepping back a bit - as you said, in the old version of the API, we could specify a custom final pull secret, and that would be passed down to the daemon. The old e2e would create an imagestream and place the secret into the final pull secret field, giving the daemon access to image. So, to rework the e2e into the new API, we have a couple of options:
- Have the e2e write to the MCO's slice of the internal registry, by removing the custom namespace for the imagestream  [[1](https://github.com/openshift/machine-config-operator/pull/4756#discussion_r1894324834)] [[2](https://github.com/openshift/machine-config-operator/blob/master/test/e2e-ocl/helpers_test.go#L78-L81)]. The existing internal registry secret sync mechanism(`syncClusterPullSecrets()`, or soon to be `syncInternalRegistryPullSecrets()`) will take care of ensuring that the daemon is able to access the final image. This is what is currently implemented and why our e2es are working :smile:
- Have the e2e update the global pull secret with the final pull/push secret of the imagestream, prior to applying the MachineOSConfig. We can then remove `syncClusterPullSecrets()` entirely. 

Are you leaning either way? I'm leaning to the first just because it is implemented and has been around for a bit. It does seem like some cycles were put into ensuring the [daemon had internal registry access](https://issues.redhat.com/browse/MCO-573) while we were in Dev/TechPreview, but maybe it is no longer a customer need? I'm happy to do the second one if you feeling strongly about that approach :+1:

yuqi-zhang replied:
Just a clarifying question, is there any difference for having that internal registry sync or not still? i.e. if we sync it still, does it imply that the user doesn't actually need to provide the final image pull secret via the global pull secret since the daemon has implicit access still?

djoshy replied:
If the user is setting `RenderedImagePushSpec` in a `MachineOSConfig` to the MCO's slice of the internal registry, then the user will not have to add anything to the global pull secret.

djoshy replied:
We discussed this in layering sync and decided to keep the existing internal registry behavior. This provides an additional option for the user in case they want to use the internal registry.

djoshy commented:
Last set of changes:
- ensured values set for `ObjectReference` will always meet API validation rules.
- renamed a few symbols for internal registry sync, namely `syncClusterPullSecrets` to `syncInternalRegistryPullSecrets`
- removed parallelization for the unit tests defined in `pkg/controller/build/osbuildcontroller_test.go` as they seem to flake at a higher rate in parallel mode

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/payload-with-prs 4.19 nightly https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/payload-with-prs 4.19 nightly blocking https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/payload 4.19 nightly blocking

yuqi-zhang commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl openshift/machine-config-operator#4756 github.com/openshift/api#2134

yuqi-zhang commented:
/payload-with-prs 4.19 nightly blocking openshift/machine-config-operator#4756 openshift/api#2134

yuqi-zhang commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl #4756 openshift/api#2134

yuqi-zhang commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl openshift/api#2134

djoshy commented:
/retest

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
This push has a fix to ensure that the baseImagePullSecret(if defined) is also checked to validate an MOSC. This was incorrectly removed when I moved the API to v1 and the field become optional. Thanks @sergiordlr !

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/api/pull/2134

djoshy commented:
/test bootstrap-unit

djoshy commented:
/test all

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl openshift/api#2134

yuqi-zhang commented:
/payload-with-prs 4.19 nightly blocking https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/retest

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/api/pull/2134

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 github.com/https://github.com/openshift/api/pull/2134

djoshy commented:
Rebased to fix merge conflicts

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/payload-with-prs 4.19 nightly blocking https://github.com/openshift/machine-config-operator/pull/4756 https://github.com/openshift/api/pull/2134

djoshy commented:
/payload-with-prs 4.19 nightly blocking https://github.com/openshift/api/pull/2134 https://github.com/openshift/machine-config-operator/pull/4756

djoshy commented:
/payload-with-prs 4.19 nightly blocking openshift/machine-config-operator#4756 openshift/api#2134

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/api/pull/2134

sergiordlr commented:
Verified with https://github.com/openshift/api/pull/2134  using IPI on AWS 

All MCOQE e2e test cases passed.

/label qe-approved

djoshy commented:
/payload-job-with-prs periodic-ci-openshift-release-master-ci-4.19-e2e-aws-ovn-upgrade openshift/machine-config-operator#4756 openshift/api#2134

djoshy commented:
/payload-job-with-prs periodic-ci-openshift-hypershift-release-4.19-periodics-e2e-aws-ovn-conformance openshift/api#2134

djoshy commented:
Rebased to fix merge conflicts!
/honk

djoshy commented:
/testwith openshift/machine-config-operator/master/e2e-gcp-op-ocl https://github.com/openshift/api/pull/2134

djoshy commented:
/retest-required

yuqi-zhang reviewed (approved):
/lgtm

Given all the testing, I think we have a solid enough foundation to get this merged. Thanks everyone for all the hard work!

deads2k commented:
merging https://github.com/openshift/api/pull/2134 and https://github.com/openshift/machine-config-operator/pull/4756 as a pair for promotion.  @yuqi-zhang and @JoelSpeed agreed they're ready.
